topology/probes/lldp/lldp.go

```diff
@@ -0,0 +1,424 @@
+		ifrHwaddr: sockaddr,
+	}
+
+	_, _, ep := unix.Syscall(unix.SYS_IOCTL, uintptr(fd),
+		unix.SIOCADDMULTI, uintptr(unsafe.Pointer(ifr)))
+
+	if ep != 0 {
+		return syscall.Errno(ep)
+	}
+	return nil
+}
+
+func (p *Probe) startCapture(ifName, mac string, n *graph.Node) error {
+	var err error
+	var handle *probes.AFPacketHandle
+
+	p.interfaceMap[ifName] = true
+
+	headerSize := uint32(lldpSnapLen)
+	fnc := func() error {
+		handle, err = probes.NewAFPacketHandle(ifName, int32(headerSize))
+		if err != nil {
+			return fmt.Errorf("Error while opening device %s: %s", ifName, err)
+		}
+		return nil
+	}
+
+	if err = common.Retry(fnc, 2, 100*time.Millisecond); err != nil {
+		return fmt.Errorf("Failed to capture packets on %s: %s", ifName, err)
+	}
+
+	lldpPrefix := "01:80:c2:00:00"
+	for _, lastByte := range []byte{0x00, 0x03, 0x0e} {
+		if err := addMulticastAddr(ifName, fmt.Sprintf("%s:%02X", lldpPrefix, lastByte)); err != nil {
+			return fmt.Errorf("Failed to add multicast address: %s", err)
+		}
+	}
+
+	packetSource := gopacket.NewPacketSource(handle, layers.LayerTypeEthernet)
+
+	bpfFilter := `(ether[0] & 1 = 1 and
+		((ether proto 0x88cc and (ether dst 01:80:c2:00:00:0e or
+								  ether dst 01:80:c2:00:00:03 or
+								  ether dst 01:80:c2:00:00:00)) or
+		 (ether dst 01:e0:52:cc:cc:cc) or
+		 (ether dst 01:00:0c:cc:cc:cc) or
+		 (ether dst 01:00:81:00:01:00))) or
+	   (ether dst 00:e0:2b:00:00:00)`
+	bpfFilter += fmt.Sprintf(" and !(ether src %s)", mac)
+
+	var rawBPF []bpf.RawInstruction
+	if rawBPF, err = flow.BPFFilterToRaw(layers.LinkTypeEthernet, headerSize, bpfFilter); err != nil {
+		handle.Close()
+		return err
+	}
+
+	if err = handle.TPacket().SetBPF(rawBPF); err != nil {
+		handle.Close()
+		return err
+	}
+
+	p.wg.Add(1)
+
+	go func() {
+		defer func() {
+			logging.GetLogger().Infof("Stopping LLDP capture on %s", ifName)
+			handle.Close()
+
+			p.Lock()
+			p.interfaceMap[ifName] = false
+			p.Unlock()
+
+			p.wg.Done()
+		}()
+
+		for atomic.LoadInt64(&p.state) == common.RunningState {
+			packet, err := packetSource.NextPacket()
+			switch err {
+			case nil:
+				lldpLayer := packet.Layer(layers.LayerTypeLinkLayerDiscovery)
+				if lldpLayer != nil {
+					lldpLayer := lldpLayer.(*layers.LinkLayerDiscovery)
+
+					chassisID := graph.GenID(string(lldpLayer.ChassisID.ID), lldpLayer.ChassisID.Subtype.String())
+					chassisMetadata := graph.Metadata{
+						"LLDP": map[string]interface{}{
+							"ChassisIDType": lldpLayer.ChassisID.Subtype.String(),
+						},
+						"Type":  "switch",
+						"Probe": "lldp",
+					}
+
+					switch lldpLayer.ChassisID.Subtype {
+					case layers.LLDPChassisIDSubTypeMACAddr:
+						common.SetField(chassisMetadata, "LLDP.ChassisID", net.HardwareAddr(lldpLayer.ChassisID.ID).String())
+					default:
+						common.SetField(chassisMetadata, "LLDP.ChassisID", string(lldpLayer.ChassisID.ID))
+					}
+
+					chassisMetadata["Name"], _ = common.GetField(chassisMetadata, "LLDP.ChassisID")
```

Comment:
maybe define L195 var chassisID string, set it L199 and 201 then SetField for Name and LLDP.ChassisID

Reply:
Done